ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -340,3 +372,165 @@ pprPatSynTypeWithoutForalls p = pprPatSynType pWithoutTypeVariables
+      bindToSig' = bindToSig hsc (tcg_rdr_env gblEnv)
+      findSigs (WhereBindings bindings _) = fmap findSig bindings
+        where findSig (WhereBinding bindingId _ _) = sequence (bindingId, bindToSig' bindingId)
+  (_, Map.fromList . fromMaybe [] -> sigMap) <-
+    initTcWithGbl hsc gblEnv ghostSpan $ sequence $ concatMap findSigs localBindings
+  pure $ Just (WhereBindingTypeSigsResult (localBindings, sigMap))
+whereBindingType _ _                   = pure Nothing
+
+-- | All where clauses from type checked source.
+findWhereQ :: GenericQ [HsLocalBinds GhcTc]
+findWhereQ = everything (<>) $ mkQ [] (pure . findWhere)
+  where
+    findWhere :: GRHSs GhcTc (LHsExpr GhcTc) -> HsLocalBinds GhcTc
+    findWhere = grhssLocalBinds
+
+-- | Find all bindings for **one** where clause.
+findBindingsQ :: GenericQ (Maybe WhereBindings)
+findBindingsQ = something (mkQ Nothing findBindings)
+  where
+    findBindings :: NHsValBindsLR GhcTc -> Maybe WhereBindings
+    findBindings (NValBinds binds sigs) =
+      Just $ WhereBindings
+        { bindings = concat $ mapMaybe (something (mkQ Nothing findBindingIds) . snd) binds
+        , existingSigNames = concatMap findSigIds sigs
+        }
+
+    findBindingIds :: LHsBindLR GhcTc GhcTc -> Maybe [WhereBinding]
+    findBindingIds bind = case unLoc bind of
+      FunBind{..} ->
+        let whereBinding = WhereBinding (unLoc fun_id) (getLoc fun_id)
+                           (col (getLoc fun_id) - col (getLoc bind))
+        in Just $ pure whereBinding
+      PatBind{..} -> Just $ (everything (<>) $ mkQ [] (fmap (uncurry wb) . maybeToList . findIdFromPat)) pat_lhs
+        where
+          wb id srcSpan = WhereBinding id srcSpan (col srcSpan - col (getLoc pat_lhs))
+      _           -> Nothing
+      where
+        col = srcSpanStartCol . realSrcSpan
+
+    -- | Example: Find `a` and `b` from @(a,b) = (1,True)@
+    findIdFromPat :: Pat GhcTc -> Maybe (Id, SrcSpan)
+    findIdFromPat (VarPat _ located) = Just (unLoc located, getLoc located)
+    findIdFromPat _                  = Nothing
+
+    findSigIds :: GenLocated l (Sig GhcRn) -> [IdP GhcRn]
+    findSigIds (L _ (TypeSig _ names _)) = map unLoc names
+    findSigIds _                         = []
+
+-- | Provide code lens for where bindings.
+whereClauseInlayHints :: PluginMethodHandler IdeState Method_TextDocumentInlayHint
+whereClauseInlayHints state plId (InlayHintParams _ (TextDocumentIdentifier uri) visibleRange)  = do
+  enabled <- liftIO $ runAction "inlayHint.config" state $ usePropertyAction #whereInlayHintOn plId properties
+  if not enabled then pure $ InL [] else do
+    nfp <- getNormalizedFilePathE uri
+    (WhereBindingTypeSigsResult (localBindings, sigMap), pm)
+        <- runActionE "InlayHint.GetWhereBindingTypeSigs" state $ useWithStaleE GetWhereBindingTypeSigs nfp
+    let bindingToInlayHints id sig = generateWhereInlayHints (T.pack $ printName (idName id)) (maybe "_" T.pack sig)
```

Comment:
I think we should probably just omit things where we don't have a signature, rather than putting in `_`?